src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ServiceDescriptionService.java

```diff
@@ -498,7 +528,178 @@ public ServiceDescriptionType refreshServiceDescription(Long id, boolean ignoreW
     }
 
     /**
-     * Get one ServiceDescriptionType by id
+     * Update Rest service description
+     *
+     * @param id
+     * @param url
+     * @param restServiceCode
+     * @param newRestServiceCode
+     * @return {@link ServiceDescriptionType}
+     * @throws UrlAlreadyExistsException
+     * @throws ServiceCodeAlreadyExistsException
+     */
+    public ServiceDescriptionType updateRestServiceDescription(Long id, String url, String restServiceCode,
+                                                               String newRestServiceCode)
+            throws UrlAlreadyExistsException, ServiceCodeAlreadyExistsException {
+        verifyAuthority("EDIT_REST");
+
+        if (newRestServiceCode == null) {
+            newRestServiceCode = restServiceCode;
+        }
+
+        ServiceDescriptionType serviceDescription = getServiceDescriptiontype(id);
+        serviceDescription.setUrl(url);
+        updateServiceCodes(restServiceCode, newRestServiceCode, serviceDescription);
+
+        checkDuplicateServiceCodes(serviceDescription);
+        checkDuplicateUrl(serviceDescription);
+
```

Comment:
Should this also check for duplicate endpoints? There is a [unique constraint index for endpoints](https://github.com/nordic-institute/X-Road/blob/c5aa6118ec6d633a28dde9d02ac1f3daa1386281/src/packages/src/xroad/common/proxy/usr/share/xroad/db/serverconf/9-rest-auth-refactoring.xml#L67) saying that there cannot be two endpoints with the same client_id & servicecode & method & path.

Duplicate endpoints should probably be checked when adding/updating either REST or OPENAPI3 ServiceDescriptions.

Reply:
This is good observation. I'm pretty sure no changes need to made because:

checkDuplicateServiceCodes makes sure there are no duplicate **service codes** for the linked **client**.

- Add rest adds always unique servicecode to client ( and **/*  method/path endpoint) -> no need to check duplicate endpoint
- Edit rest doesn't touch method/path and if it changes servicecode it's uniqueness is ensured in checkDuplicateServiceCodes
- Add openapi adds unique servicecode to client and 1-n endpoints but it shouldn't be necessary to handle duplicate endpoints as parsing openapi description will throw ParsingException in that case. I added a unit test for that.
- Edit openapi3 changes the servicecode and/or endpoints but yet again there will not be duplicate servicecode and if an existing, manually added endpoint, is equivalent (See EndpointType.java changes in this PR) to one in openapi3 description it will be changed to generated = true. Also again there can't be duplicate endpoints in the openapi3 description or ParsingException is thrown

Reply:
As it turned out this was caused by Spring **not** by default rolling back transactions if a checked exception was thrown. This has also been fixed now